### ISSUE TITLE
[Bug]: During Upgrade - PHP Fatal Error - Not compatible Declaration of RoomShareProvider::getSharesInFolder

### ISSUE THREAD

**Issue by hendrikreimers** [nextcloud/server#34673]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Tried to upgrade my latest Nextcloud, version 24 to the current 25.0.0 release.
After running ./occ upgrade it crashes with an PHP Fatal error:

> PHP Fatal error:  Declaration of OCA\Talk\Share\RoomShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares): array must be compatible with OCP\Share\IShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares, $shallow = true) 

Deactivated all incompatible apps before startet upgrade.

### Steps to reproduce

1. Upgrade from NC 24 to 25 in browser
3. run ./occ upgrade on console as recommended

### Expected behavior

No fatal error upgrade finishing slightly

### Installation method

Community Manual installation with Archive

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Apache (supported)

### Database engine version

SQlite

### Is this bug present after an update or on a fresh install?

Updated to a major version (ex. 22.2.3 to 23.0.1)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "mydomain.tld"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "sqlite3",
        "version": "24.0.6.1",
        "overwrite.cli.url": "https:\/\/mydomain.tld\/",
        "installed": true,
        "loglevel": 3,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "remember_login_cookie_lifetime": 86400,
        "session_lifetime": 86400,
        "auth.bruteforce.protection.enabled": true,
        "versions_retention_obligation": "auto, 30",
        "default_language": "de",
        "default_locale": "de_DE",
        "knowledgebaseenabled": false,
        "htaccess.RewriteBase": "\/",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauth": 1,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "overwritehost": "mydomain.tld",
        "maintenance": false,
        "theme": "",
        "data-fingerprint": "***REMOVED SENSITIVE VALUE***",
        "app_install_overwrite": [
            "keeweb",
            "bruteforcesettings",
            "ransomware_protection",
            "onlyoffice",
            "documentserver_community",
            "files_antivirus"
        ],
        "integrity.check.disabled": true,
        "default_phone_region": "DE",
        "3rdpartyroot": "\/var\/www\/nextcloud\/public_html\/"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - bruteforcesettings: 2.4.0
  - calendar: 3.5.0
  - circles: 24.0.1
  - cloud_federation_api: 1.7.0
  - contacts: 4.2.2
  - dav: 1.22.0
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_accesscontrol: 1.14.1
  - files_antivirus: 3.3.1
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - firstrunwizard: 2.13.0
  - keeweb: 0.6.9
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.1
  - oauth2: 1.12.0
  - password_policy: 1.14.0
  - photos: 1.6.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.5
  - tasks: 0.14.4
  - text: 3.5.1
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - updatenotification: 1.14.0
  - viewer: 1.8.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit
  - comments: 1.10.0
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - encryption
  - files_external
  - ransomware_protection: 1.13.0
  - support: 1.1.0
  - survey_client: 1.6.0
  - systemtags: 1.10.0
  - user_ldap
  - user_status: 1.0.0
  - weather_status: 1.0.0
```

### Nextcloud Signing status

```shell
Integrity checker has been disabled. Integrity cannot be verified.
```

**Comment by szaimen** [comment]
Thank you for taking the time to report a bug! 👍

As this seems to be a setup issue I would like to ask you to raise your question in the forums: https://help.nextcloud.com

**Comment by nickvergessen** [comment]
Can you please post the full log line from the nextcloud.log file? That would really be helpful

**Comment by regs01** [comment]
```
{"reqId":"B7Oq4izcb98xOhDwhjq2","level":3,"time":"2022-10-20T01:27:24+00:00","remoteAddr":"<IP>","user":"--","app":"PHP","method":"GET","url":"/apps/files/","message":"Declaration of OCA\\Talk\\Share\\RoomShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares): array must be compatible with OCP\\Share\\IShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares, $shallow = true) at /var/www/nextcloud/html/apps/spreed/lib/Share/RoomShareProvider.php#520","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0","version":"24.0.6.1","data":{"app":"PHP"}}
```

Looks like even fails to upgrade. But now it's just white page.

**Comment by robjuffermans** [comment]
I'm using the docker version and had the same problem, white screen after login.

I solved this using: `docker-compose exec --user www-data app php occ app:update --all`

**Comment by wuan** [comment]
Had the describe problem when updating a non docker version. Adding the additional last argument `$shallow = true` to the function `getSharedInFolder`
in `apps/spreed/lib/Share/RoomShareProvider.php` on line 520. After that, the upgrade to 25.0.0 could be continued without problems.

**Comment by VincentSC** [comment]
Our update script now runs the updates twice, as it failed to work when doing just once:

```
echo Updating to latest version
sudo docker-compose pull && sudo docker-compose up -d
sleep 2
./occ.sh upgrade
./occ.sh app:update --all
./occ.sh upgrade
./occ.sh app:update --all
./occ.sh maintenance:mode --off
```

occ.sh being:
```
#!/bin/bash
docker exec -it -u 33 nextcloud_app /var/www/html/occ $@

# For help, see: https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/occ_command.html
```

**Comment by szaimen** [comment]
The problem will get fixed with 25.0.1

**Comment by eplanet** [comment]
In the meantime I'd love that you do not propose upgrades to `25.0.0` because I was just caught into it and can't fix so easily. It would be nice to avoid further people getting trapped in this bug.

**Comment by VincentSC** [comment]
Credits go to @robjuffermans https://github.com/nextcloud/server/issues/34673#issuecomment-1284885880

I just like documentation and automation, so the only thing I added were these.

**Comment by eplanet** [comment]
For the record in my case it's `occ upgrade` that unblocked the situation, not the `occ app:update --all`

**Comment by hendrikreimers** [comment]
It also helps if you disabled the "Talk" App before upgrade to NC 25 and after upgrade enable it again.

**Comment by rheinFrohnatur** [comment]
> It also helps if you disabled the "Talk" App before upgrade to NC 25 and after upgrade enable it again.

I ran into this error while updating to 25.0.1 via the web based updater. Now I only get a white screen, nothing else. Is there a way to disable the "Talk" app directly in the database? I could not find it in table `oc_appconfig`.

I have been running NC in a shared hosting environment for years now and newer had any trouble updating. Unfortunately I cannot run scripts on this server so running `occ` is not an option for me.

Any suggestions?

**Comment by rheinFrohnatur** [comment]
In the end I think I fixed my problem in a very dirty way. I just altered

`public function getSharesInFolder($userId, Folder $node, $reshares): array {` to

`public function getSharesInFolder($userId, Folder $node, $reshares, $shallow = true): array {`

in `RoomShareProvider.php`. Now NC 25 is running again. Maybe this helps others stumbling upon this error in the future.

**Comment by gohrner** [comment]
@szaimen 

> The problem will get fixed with 25.0.1

No, it didn't - upgrading to 25.0.1 with the web updater just nuked my instance.

Very annoying that an update is release with such a severe bug.

Will now try to resurrect it using the command line updater... :-/

**Comment by vladas-palubinskas** [comment]
The same issue for me too, upgrading from 24.0.7 to 25.0.1 via Webupdater — the installation finished OK, but a WWW browser started as a blank screen. FreeBSD 13.1-RELEASE-p3 on amd64, PHP 8.1.12, Hiawatha webserver 11.2. From Nextcloud error.log:

`/usr/local/www/nextcloud/index.php|PHP message: PHP Fatal error:  Declaration of OCA\Talk\Share\RoomShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares): array must be compatible with OCP\Share\IShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares, $shallow = true) in /usr/local/www/nextcloud/apps/spreed/lib/Share/RoomShareProvider.php on line 520`
`CGI returned 500 Internal Error`

**Comment by vladas-palubinskas** [comment]
I followed @rheinFrohnatur trick and upgraded 24.0.7 to 25.0.1 successfully, with some restrictions, though:

> The following apps have been disabled:
> — accessibility (incompatible);
> — files_videoplayer (incompatible);
> — spreed (incompatible);
> — talk_matterbridge (incompatible).

![20221119nxc](https://user-images.githubusercontent.com/44341097/202863706-8d29f90b-916e-41f2-8c72-b9cfb891a16b.png)

**Comment by jeffh2os** [comment]
Same issue & symptoms on upgrade from 24.0.7 to 25.0.1 via web updater (vmware install on ubuntu) reverted to snapshot until we have a fix for this. 
`PHP Fatal error:  Declaration of OCA\\Talk\\Share\\RoomShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares): array must be compatible with OCP\\Share\\IShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares, $shallow = true) in /var/www/html/apps/spreed/lib/Share/RoomShareProvider.php on line 520`

**Comment by didyouexpectthat** [comment]
`occ upgrade` fixed this for me without disabling Talk. But it failed using the UI.

**Comment by jajaX** [comment]
Hi,
same problem since upgrade to 25.01.

Talk is disable now and I can't activate it.

Declaration of OCA\Talk\Share\RoomShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares): array must be compatible with OCP\Share\IShareProvider::getSharesInFolder($userId, OCP\Files\Folder $node, $reshares, $shallow = true) at /var/www/nextcloud/apps/spreed/lib/Share/RoomShareProvider.php#520

**Comment by KlausHans** [comment]
I wonder why this is closed. The issue persists! The upgrade failed to for me from 24.0.7 to 25.0.1 via gui. I really had to change some php code by hand, wow.

**Comment by gohrner** [comment]
> I wonder why this is closed. The issue persists! The upgrade failed to for me from 24.0.7 to 25.0.1 via gui.

This issue apparently was about a bug in Spreed / Talk, which was fixed, but which unfortunately did not fix the upgrade. Therefore I created the new issue https://github.com/nextcloud/server/issues/35289 two weeks ago, as I doubted that a closed issues will still receive much attention.

Unfortunately, the new issue and the upgrade problem are still not resolved yet.

**Comment by gremaldi** [comment]
Same issue from 24.0.8 to 25.0.2. via Webupdater today
PHP 7.4

I use the hack from [rheinFrohnatur](https://github.com/rheinFrohnatur). Thanks for that.

**Comment by ynad** [comment]
> 

same problem here, PHP 7.4 and upgrade from 24.0.8 to 25.0.2. Solved as suggested above.

**Comment by ikonnow** [comment]
> In the end I think I fixed my problem in a very dirty way. I just altered
> 
> `public function getSharesInFolder($userId, Folder $node, $reshares): array {` to
> 
> `public function getSharesInFolder($userId, Folder $node, $reshares, $shallow = true): array {`
> 
> in `RoomShareProvider.php`. Now NC 25 is running again. Maybe this helps others stumbling upon this error in the future.

the issue was not fixed with the 25.0.1 - and nothing from the above except this one here worked!
To alter the file went with ash _container-name_ - found the .php file and altered it with vi. Yes dirty but sure works. Took some time!

**Comment by nud** [comment]
I just ran into this issue when upgrading from 24.0.8 to 25.0.2, so I'd say it's not fixed either.

Upgrading from CLI worked.

**Comment by BernhardPosselt** [comment]
Not fixed in 25.0.3

**Comment by lightweight** [comment]
Affected me just now with upgrade (one of 4 instances) to 26.0.0.11... applied shown in https://github.com/nextcloud/server/issues/34673#issuecomment-1374284147 but still getting an error due to "app view (viewer) not replaced correctly"... Will report back if I solve it. (using Docker install)

**Comment by VincentSC** [comment]
Could it be the password-app again? This app has caused quite some problems in the past.

Below is the current version of my `update.sh`. Just see what I need to do, to make sure the update gets 100% every time:
- a pause of 2 seconds
- disabling passwords app
- running update twice
- repairing the database

It would be better if this is handled by NC itself, and we only need to do a pull. I also would like to be able to update the apps while in maintenance mode.
```
#! /bin/bash

echo Updating Nextcloud to latest version

# historically the passwords app has caused the most problems,
# and it has proven to be effective to disable the app during upgrades
./occ.sh app:disable passwords

./occ.sh maintenance:mode --on
sudo docker-compose pull && sudo docker-compose up -d
# No joke: the occ upgrade/update commands don't work in maintenance mode....

# prevent a "Failed to connect to the database" error
sleep 2

./occ.sh maintenance:mode --off

# update all twice, as it sometimes forgets small details the first time
./occ.sh upgrade
./occ.sh app:update --all
./occ.sh app:enable passwords
./occ.sh upgrade
./occ.sh app:update --all
./occ.sh maintenance:repair
```

Note: I am experimenting with the enabling of passwords inbetween the two app-updates. I used to do this at the very end. Also the repair-step at the end I doubt if I really need that.

**Comment by petajoulecorp** [comment]
How is this still a thing?

update 24.0.12.something to 25.0.9 failed

`[Mon Jul 24 13:57:00.629976 2023] [php7:error] [pid 1238002] [client xxxx] PHP Fatal error:  Declaration of OCA\\Talk\\Share\\RoomShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares): array must be compatible with OCP\\Share\\IShareProvider::getSharesInFolder($userId, OCP\\Files\\Folder $node, $reshares, $shallow = true) in /var/www/nextcloud/apps/spreed/lib/Share/RoomShareProvider.php on line 520`

After adding "$shallow = true" as last parameter in line 520 the update could continue. Thx @wuan 

The Date is Monday, 24 July 2023.

**Comment by VincentSC** [comment]
Ah, the Talk-app. I turned it off some time ago, as it caused more problems. A `./occ.sh app:disable talk` might also work - see script above for details.

This is my occ.sh:
```
#!/bin/bash
docker exec -it -u 33 app_nextcloud /var/www/html/occ $@

# For help, see: https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/occ_command.html
```
You might need to specify a different user ID for `www-data`

**Comment by VincentSC** [comment]
The main problem is that the update script of apps and core are not in sync. A very simple solution is that an update requires all apps to be updated before an upgrade. My script enforces it.
